mlir/lib/Catalyst/Transforms/SplitMultipleTapes.cpp

```diff
@@ -0,0 +1,307 @@
+                // reached post processing
+                cur_tape++;
+            }
+        });
+    } // collectOperationsForEachTape()
+
+    void collectNecessaryValuesFromEarlierTapes(const SmallVector<Operation *> &TapeOps,
+                                                SmallVector<Value> &NecessaryValuesFromEarlierTapes)
+    {
+        // Go through a list of operations and collect all the necessary operand values
+        // not defined in this list itself, and add them to NecessaryValuesFromEarlierTapes
+        for (Operation *PPOp : TapeOps) {
+            PPOp->walk([&](Operation *op) {
+                for (auto operand : op->getOperands()) {
+                    Operation *OperandSource = operand.getDefiningOp();
+                    if (std::find(TapeOps.begin(), TapeOps.end(), OperandSource) == TapeOps.end()) {
+                        // A list operand not produced in list itself, must be from earlier
+                        // tapes/preprocessing!
+                        if (std::find(NecessaryValuesFromEarlierTapes.begin(),
+                                      NecessaryValuesFromEarlierTapes.end(),
+                                      operand) == NecessaryValuesFromEarlierTapes.end()) {
+                            NecessaryValuesFromEarlierTapes.push_back(operand);
+                        }
+                    }
+                }
+            });
+        }
+    } // collectNecessaryValuesFromEarlierTapes()
+
+    void getNecessaryTapeReturns(SmallVector<Value> &RetValues,
+                                 std::shared_ptr<SmallVector<Operation *>> TapeOps,
+                                 SmallVector<Value> &NecessaryValuesFromEarlierTapes)
+    {
+        for (Value v : NecessaryValuesFromEarlierTapes) {
+            Operation *VDefOp = v.getDefiningOp();
+            if (std::find(TapeOps->begin(), TapeOps->end(), VDefOp) != TapeOps->end()) {
+                // This Value needed for PP is in this tape!
+                RetValues.push_back(v);
+            }
+        }
+
+        // We process the tapes in reverse order
+        // Hence from the viewpoint of the earlier tapes,
+        // a processed later tape (which has become a call op) is also their "post processing"
+        // The later tape's arguments need to be returned by the earlier tapes
+        // Hence when processing a tape, it needs to ask for earlier tapes to return
+        // values it needs, just like how post processing asks for all tapes to return
+        // values it needs.
+        collectNecessaryValuesFromEarlierTapes(*TapeOps, NecessaryValuesFromEarlierTapes);
+    } // getNecessaryTapeReturns()
+
+    std::pair<scf::ExecuteRegionOp, scf::YieldOp> wrapTapeOpsInSCFRegion(
+        std::shared_ptr<SmallVector<Operation *>> TapeOps, const SmallVector<Value> &RetValues,
+        const SmallVector<mlir::Type> &RetTypes, IRRewriter &builder, Location loc)
+    {
+        builder.setInsertionPoint(TapeOps->front());
```

Comment:
I would use [`InsertionGuard`s](https://mlir.llvm.org/doxygen/classmlir_1_1OpBuilder_1_1InsertionGuard.html) as opposed to `setInsertionPoint`

Reply:
The three insertion points here actually need to be explicitly set; I can add an insertion guard before them, if that's what you mean